### PR TITLE
[COD-22] Create group [Backend]

### PR DESCRIPTION
**Pull request by LorDEvilDead**
[legolas.atlassian.net/jira/software/projects/COD/boards/1?selectedIssue=COD-22](url)
Create group when send request post to /api/v1/group/ and return code: 201. With correct parameters ({name: String}). Then new group entity is created in database. And it attached to the current user. And with incorrect parameters ({name:Nill]) group doesn't create.

### CODE REVIEW

File: backend/.env 2 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+PORT=3000
```
Comment (npupko) [inline backend/.env 2:1]: Do we need this file?

File: backend/.env 3 [hunk 2]
```diff
@@ -0,0 +1,2 @@
+PORT=3000
```
Comment (npupko) [inline backend/.env 3:1]: Do we need this file?

File: backend/app/controllers/api/v1/groups_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  module V1
+    class GroupsController < ApplicationController
+      include Dry::Monads::Result::Mixin
+
+      def create
+        case group_creation
+        in Success(group)
+          render json: group, status: :created
+        in Failure(error)
+          render json: error, status: :unprocessable_entity
```
Comment (npupko) [inline backend/app/controllers/api/v1/groups_controller.rb:13]: You're returning errors here. But information about this you have only here for now.
We need to explicitly tell the frontend what this array we're getting from the backend here is. You could do it via passing a specific key. `errors` in this case.

```ruby
render json: error, status: :unprocessable_entity
```

You could check the [ticket description](https://legolas.atlassian.net/browse/COD-22) more closely and find out by yourself:

> And body of response contains {errors: []}

File: backend/app/models/group.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class Group < ApplicationRecord
+  belongs_to :user
+  validates :name, length: { minimum: 4, maximum: 20 }
```
Comment (npupko) [inline backend/app/models/group.rb:5]: Why do we want to have `maximum: 20` limit there? Let's not limit users without specific reasons

File: backend/config/routes.rb [hunk 5]
```diff
@@ -16,6 +16,9 @@
       end
       post '/refresh', to: 'sessions#refresh'
       post 'hello', to: 'sessions#hello'
+      resources :groups do
```
Comment (npupko) [inline backend/config/routes.rb:19]: This is not correct. Feel free to check [this article](https://guides.rubyonrails.org/routing.html) and fix it

File: backend/app/models/group.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class Group < ApplicationRecord
+  belongs_to :user
```
Comment (npupko) [inline backend/app/models/group.rb:4]: Let's name `user` as `author` here, because we will have students, also attached to the group and we will need to determine who is who:)

Expected result:

```ruby
Group.last.author #=> <User #123 ...>
```

File: backend/app/services/groups/create.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module Groups
+  class Create < Core::Service
+    def call(params:, user:)
+      @params = params
+      @user = user
+      group.user_id = user.id
+      group.save ? Success(group) : Failure(group.errors)
```
Comment (npupko) [inline backend/app/services/groups/create.rb:9]: Let's move it (and line above) to the separate `create_group` method and use `yield create_group` there

File: backend/db/migrate/20230129163956_create_groups.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateGroups < ActiveRecord::Migration[7.0]
+  def change
+    create_table :groups do |t|
+      t.string :name
+      t.integer :user_id
```
Comment (npupko) [inline backend/db/migrate/20230129163956_create_groups.rb:7]: Let's use `references` with `foreign_key here or create index manually ([docs](https://edgeguides.rubyonrails.org/active_record_migrations.html#creating-associations-using-references))

File: backend/db/schema.rb [hunk 9]
```diff
@@ -28,6 +27,23 @@
...
   end
 
+
+  create_table "courses", force: :cascade do |t|
+    t.string "name"
+    t.bigint "user_id", null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["user_id"], name: "index_courses_on_user_id"
+  end
```
Comment (npupko) [inline backend/db/schema.rb:37]: This is not related to your PR, right?

File: backend/spec/factories/groups.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :group do
+    name { 'MyString' }
+    user_id { 1 }
```
Comment (npupko) [inline backend/spec/factories/groups.rb:6]: Let's create user here directly ([docs](https://www.rubydoc.info/gems/factory_bot/file/GETTING_STARTED.md#associations))

File: backend/spec/helpers/api/v1/groups_helper_spec.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
```
Comment (npupko) [inline backend/spec/helpers/api/v1/groups_helper_spec.rb:1]: We don't need empty file there

File: backend/spec/models/group_spec.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  let(:current_user) { create(:user) }
+  let(:name) { 'SAWQ' }
+  let(:user_id) { current_user.id }
+  let(:group_params) { { name:, user_id: } }
+
+  it 'creates a group' do
+    expect(group1).to be_persisted
+  end
+
+  it 'realy creates a group' do
```
Comment (npupko) [inline backend/spec/models/group_spec.rb:17]: The name of this test is not clear enough. You could go with something like "increments amount of records in the database by one"

**Review by npupko — CHANGES_REQUESTED** [review]
Also, please, add all the `.DS_Store` files to `.gitignore` and remove them from index

File: backend/db/schema.rb (hunk 9 above)
Comment (LorDEvilDead) [reply backend/db/schema.rb:37]: yep. i just merged it, because i had conflict with it after pulling.

File: backend/app/models/group.rb (hunk 4 above)
Comment (LorDEvilDead) [reply backend/app/models/group.rb:5]: I thought that will be owfull if name of group will something like that: **"Ruby On Rails Group with my friends and my grand ma (granny)"**

**Comment by Set27** [comment]
Check CI

File: backend/config/routes.rb [hunk 13]
```diff
@@ -16,6 +16,8 @@
       end
       post '/refresh', to: 'sessions#refresh'
       resources :courses, only: [:create]
+      post 'hello', to: 'sessions#hello'
```
Comment (Set27) [inline backend/config/routes.rb:19]: Do we need it here?

File: backend/spec/factories/groups.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :group do
+    author_last_name { 'SAWQ' }
```
Comment (Set27) [inline backend/spec/factories/groups.rb:5]: Make author last name valid (not Upper case)

File: backend/spec/models/group_spec.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Group do
+  subject(:group1) { described_class.create(group_params) }
+
+  let(:current_user) { create(:user) }
+  let(:name) { 'SAWQ' }
```
Comment (Set27) [inline backend/spec/models/group_spec.rb:9]: make name valid

File: backend/spec/requests/api/v1/groups_spec.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'Api::V1::Groups' do
+  describe 'post /create'
+  let(:user) { create(:user) }
+  let(:name) { 'SAWQ' }
```
Comment (Set27) [inline backend/spec/requests/api/v1/groups_spec.rb:8]: up

File: backend/config/routes.rb (hunk 13 above)
Comment (LorDEvilDead) [reply backend/config/routes.rb:19]: > Do we need it here?
It's not mine

File: backend/spec/factories/groups.rb (hunk 14 above)
Comment (LorDEvilDead) [reply backend/spec/factories/groups.rb:5]: > Make author last name valid (not Upper case)

It's not make sense

File: backend/spec/models/group_spec.rb (hunk 15 above)
Comment (LorDEvilDead) [reply backend/spec/models/group_spec.rb:9]: > make name valid

It's not make sense
There i check minimum symbols in my model

File: backend/spec/requests/api/v1/groups_spec.rb (hunk 16 above)
Comment (LorDEvilDead) [reply backend/spec/requests/api/v1/groups_spec.rb:8]: > up

up

**Comment by LorDEvilDead** [comment]
> Also, please, add all the `.DS_Store` files to `.gitignore` and remove them from index

Forget about that. I will do it today

File: backend/config/routes.rb [hunk 17]
```diff
@@ -13,6 +13,7 @@
       end
       post '/refresh', to: 'sessions#refresh'
       resources :courses, only: [:create]
+      resources :groups, only: %i[create show update destroy]
```
Comment (MikhailZaitsau) [inline backend/config/routes.rb:16]: I'm not sure... Do we need any routes except "create" now?

File: backend/spec/factories/groups.rb (hunk 10 above)
Comment (npupko) [reply backend/spec/factories/groups.rb:8]: This factory is not working. Please, make sure it creates group correctly
```ruby
FactoryBot.create(:group)
```

File: backend/app/models/user.rb [hunk 18]
```diff
@@ -12,4 +12,5 @@ class User < ApplicationRecord
   has_many :whitelisted_tokens, dependent: :delete_all
   has_many :blacklisted_tokens, dependent: :delete_all
   has_many :courses, dependent: :delete_all
+  has_many :groups, dependent: :delete_all
```
Comment (npupko) [inline backend/app/models/user.rb:15]: The purpose of this line is to create methods for accessing groups via user objects:
```
User.last.groups
```
But this is not working now
image

File: backend/app/services/groups/create.rb [hunk 19]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module Groups
+  class Create < Core::Service
+    def call(params:, user:)
+      @params = params
+      @user = user
+      group = yield create_group
+      Success(group)
```
Comment (npupko) [inline backend/app/services/groups/create.rb:9]: No need to do `yield` and then wrap result again

The purpose of `yield` is to unwrap the Result if it is Success or bypass all the future logic if the result is Failure.
You could achieve the same results only just by calling `create_group` here

File: backend/db/schema.rb [hunk 20]
```diff
@@ -54,10 +62,10 @@
     t.datetime "remember_created_at"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
-    t.date "birthdate"
-    t.datetime "token_issued_at"
     t.string "last_name", default: "", null: false
     t.enum "role", null: false, enum_type: "users_role"
+    t.date "birthdate"
+    t.datetime "token_issued_at"
```
Comment (npupko) [inline backend/db/schema.rb:68]: Do we need to move these lines?

File: backend/spec/requests/api/v1/groups_spec.rb [hunk 21]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'Api::V1::Groups' do
+  describe 'post /create'
```
Comment (npupko) [inline backend/spec/requests/api/v1/groups_spec.rb:6]: What's the purpose of this line here?

**Review by npupko — CHANGES_REQUESTED** [review]

File: backend/config/routes.rb [hunk 22]
```diff
@@ -15,6 +15,8 @@
       resources :courses, only: [:create] do
         resources :lessons, only: [:create]
       end
+      resources :courses, only: [:create]
```
Comment (MikhailZaitsau) [inline backend/config/routes.rb:18]: This line duplicate line 15. Remove it please.

File: backend/app/models/user.rb [hunk 23]
```diff
@@ -13,4 +13,5 @@ class User < ApplicationRecord
   has_many :blacklisted_tokens, dependent: :delete_all
   has_many :courses, dependent: :delete_all
   has_many :lessons, through: :courses
+  has_many :groups, dependent: :delete_all, foreign_key: 'author_id', inverse_of: false
```
Comment (LorDEvilDead) [inline backend/app/models/user.rb:16]: Тут я переделаю на inverse_of: :author

**Review by npupko — APPROVED** [review]
LGTM

**Review by MikhailZaitsau — APPROVED** [review]
LGTM!
